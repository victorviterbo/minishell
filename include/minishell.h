/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   minishell.h                                        :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: vviterbo <vviterbo@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2025/01/17 20:04:30 by vviterbo          #+#    #+#             */
/*   Updated: 2025/02/06 20:26:40 by vviterbo         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#ifndef MINISHELL_H
# define MINISHELL_H

# include <stdio.h>
# include <stdlib.h>
# include <sys/stat.h>
# include <sys/wait.h>
# include <unistd.h>
# include <errno.h>
# include <limits.h>
# include <stdio.h>

# include "libft.h"

# define SHELL_NAME "Minishell: "
# define SHELL_PROMPT "Minishell$ "
# define DEFAULT_ERROR "an unknown error has occured"

typedef struct s_data
{
	t_list	**envp;
	char	**env_arr;
	int		exit_status;
}	t_data;

typedef struct s_var
{
	char	*name;
	char	*value;
}	t_var;

//exec/echo.c
int		ft_echo(char *str, bool nl);
//exec/cd.c
int		ft_cd(t_data *data, char *path);
char	*get_absolute_path(char *path);
//exec/env.c
int		ft_env(t_data *data);
//exec/export.c
int		ft_export(t_data *env, char *args[]);
//exec/pwd.c
int		ft_pwd(void);
char	*ft_get_current_path(void);
//exec/unset.c
int		ft_unset(t_data *data, char **varnames);
int		pop_var(t_data *data, char *varname);
void	free_var(void *data);
//exec/env.c
int		ft_env(t_data *data);
//exec/execve.c
int		ft_execve(t_data *data, char **args);
char	*find_exec(char *path_list, char *exec);
//utils/error_handlings.c
void	ft_print_error(const char *message);
//utils/parsing_utils.c
bool	*is_quoted(char *str, char open_char, char close_char);
//utils/variables.c
int		init_env(t_data *data, char **envp);
int		new_var(t_data *data, char *str);
int		add_var(t_data *data, t_list **env, char *str, size_t name_len);
int		change_var(t_data *data, t_list *current, char *first_equal,
			bool append);
char	*get_var(t_data *data, char *varname);
char	*get_var(t_data *data, char *varname);
//utils/env_to_arr.c
char	**env_to_arr(t_data *data);
char	*var_to_str(t_list *current);
void	update_env_arr(t_data *data);

#endif