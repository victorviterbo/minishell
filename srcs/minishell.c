/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   minishell.c                                        :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: vbronov <vbronov@student.42lausanne.ch>    +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2025/02/03 08:35:11 by vviterbo          #+#    #+#             */
/*   Updated: 2025/04/27 03:37:58 by vbronov          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "minishell.h"

enum e_signal	g_signal;

/**
 * @brief Handles incoming signals for minishell
 *
 * SIGINT: CTRL+C, SIGQUIT: CTRL+\ and CTRL+D (EOF) are handled
 * acording to the current mode of the shell,
 * either interactive (READLINE_MODE) or execution (EXECUTION_MODE).
 *
 * @param signum The signal number to be handled
 *
 * @return void
 */
void	signal_handler(int signum)
{
	if (signum == SIGINT
		&& (g_signal == READLINE_MODE || g_signal == GOT_SIGINT))
	{
		g_signal = GOT_SIGINT;
		rl_replace_line("", 0);
		ft_printf("\n");
		rl_on_new_line();
		rl_redisplay();
	}
}

static int	init_data(t_data *data, char **envp)
{
	if (save_std_streams(data) != EXIT_SUCCESS)
		return (data->exit_status);
	data->env_arr = NULL;
	data->envp = NULL;
	data->tokens = NULL;
	data->exit_status = EXIT_SUCCESS;
	data->last_exit_status = EXIT_SUCCESS;
	data->tree = NULL;
	init_env(data, envp);
	init_builtin(data);
	return (data->exit_status);
}

static void	process_line(t_data *data, char *line)
{
	lexer(data, line);
	if (data->tokens)
	{
		data->last_exit_status = data->exit_status;
		data->exit_status = EXIT_SUCCESS;
		add_history(line);
		if (DEBUG)
			print_tokens(data->tokens);
		if (syntax_check(data, data->tokens) != EXIT_SUCCESS)
			return ;
		data->tree = build_tree(data, data->tokens, NULL);
		prepare_heredocs(data, data->tree);
		if (DEBUG)
			display_tree(data->tree);
		if (data->exit_status == EXIT_SUCCESS)
			ft_run_ast(data, data->tree);
	}
}

static void	main_loop(t_data *data)
{
	char	*line;

	while (TRUE)
	{
		g_signal = READLINE_MODE;
		line = readline(SHELL_PROMPT);
		if (g_signal == GOT_SIGINT)
		{
			data->exit_status = 130;
			data->last_exit_status = data->exit_status;
		}
		g_signal = EXECUTION_MODE;
		if (!line)
			ft_exit(data, NULL, 0);
		process_line(data, line);
		free(line);
		free_tree(data->tree);
		data->tree = NULL;
		free_tokens(data->tokens);
		data->tokens = NULL;
	}
}

int	main(int argc, char **argv, char **envp)
{
	t_data				data;

	(void)argc;
	(void)argv;
	set_signal(SIGINT, signal_handler);
	set_signal(SIGQUIT, SIG_IGN);
	if (init_data(&data, envp) != EXIT_SUCCESS)
		return (data.exit_status);
	main_loop(&data);
	return (EXIT_SUCCESS);
}
