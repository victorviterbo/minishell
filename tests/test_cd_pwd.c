/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   test_cd_pwd.c                                      :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: vviterbo <vviterbo@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2025/01/15 16:02:46 by vviterbo          #+#    #+#             */
/*   Updated: 2025/01/28 18:03:07 by vviterbo         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "minishell.h"

char	**ft_make_test_strarr(char *str);

int	main(int argc, char *argv[], char *envp[])
{
	char	*here;
	t_data	*data;
	pid_t	pid;

	(void)argc;
	(void)argv;
	here = ft_get_current_path();
	data = ft_calloc(1, sizeof(t_data));
	init_env(data, envp);
	ft_pwd();
	if (ft_cd(data, "testdir") != 0)
		return (EXIT_FAILURE);
	ft_pwd();
	ft_cd(data, here);
	pid = fork();
	if (pid == 0)
		if (ft_cd(data, "testdir_readonly") == 0)
			return (EXIT_FAILURE);
	pid = fork();
	if (pid == 0)
		if (ft_cd(data, "testdir_no_right") == 0)
			return (EXIT_FAILURE);
	pid = fork();
	if (pid == 0)
		if (ft_cd(data, "test_cd_pwd.c") == 0)
			return (EXIT_FAILURE);
	pid = fork();
	if (pid == 0)
		if (ft_cd(data, "lalala") == 0)
			return (EXIT_FAILURE);
	if (ft_cd(data, "/Users/vviterbo/Desktop/42/minishell") != 0)
		return (EXIT_FAILURE);
	if (ft_cd(data, "/Users/vviterbo/Desktop/42/minishell/") != 0)
		return (EXIT_FAILURE);
	pid = fork();
	if (pid == 0)
		if (ft_cd(data, "/Users/vviterbo/Desktop/42/not_a_folder") == 0)
			return (EXIT_FAILURE);
	pid = fork();
	if (pid == 0)
		if (ft_cd(data, "/Users/vviterbo/Desktop/42/not_a_folder/") == 0)
			return (EXIT_FAILURE);
	free(here);
	ft_lstclear(data->envp, free_var);
	free(data->envp);
	free(data);
	return (EXIT_SUCCESS);
}
