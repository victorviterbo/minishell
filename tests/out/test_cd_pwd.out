/Users/vviterbo/Desktop/42/minishell/tests
/Users/vviterbo/Desktop/42/minishell/tests/testdir
Minishell: cd: testdir_readonly: Permission denied
Minishell: cd: testdir_no_right: Permission denied
Minishell: cd: test_cd_pwd.c: Not a directory
Minishell: cd: lalala: No such file or directory
Minishell: cd: ~/Desktop/42/not_a_folder: No such file or directory
Minishell: cd: ~/Desktop/42/not_a_folder/: No such file or directory
